### PR TITLE
feat(packagist): Support for `v2` protocol

### LINKED ISSUE #19495: Default composer/packagist to use v2 datasource API
### What would you like Renovate to be able to do?

When identifying the metadata of a composer repository, renovate rightfully [looks for the repository index file](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/packagist/index.ts#L61), packages.json. It then parses the packages.json, explicitly looking for composer v1 metadata keys like `packages`, `includes`, `providers-url`, `providers-lazy-url`, `provider-includes`, and `providers`.

However, composer v2 is able to gather any package metadata using the [`metadata-url` property](https://github.com/composer/composer/blob/main/src/Composer/Repository/ComposerRepository.php#L1206-L1210) of the packages.json index file, and can get the names of all packages in the repo with the [`list` property](https://github.com/composer/composer/blob/main/src/Composer/Repository/ComposerRepository.php#L448-L458). The `metadata-url` property can be used much like the `providers-lazy-url` property.  Although it looks like Renovate supports composer v2, it does not seem to have support for these metadata properties. We would like Renovate to support these index file metadata properties, as our private v2 composer repository is currently considered "empty" by Renovate.  

Just FYI - for our case, we have a private composer repository in Artifactory, and our v2 **packages.json** index file only exposes the following properties: `search`, `metadata-url`, and `list`. Our packages.json looks like this: 

``` 
{
  "search" : "https://artifactory.com/artifactory/api/composer/local-php-repository/search.json?q=%query%&type=%type%",
  "metadata-url" : "[https://artifactory.com/artifactory/api/composer/local-php-repository/p2/%package%.json",
  "list" : "https://artifactory.com/artifactory/api/composer/local-php-repository/packages/list.json"
}
```

### If you have any ideas on how this should be implemented, please tell us here.

I believe updating the [getRegistryMeta](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/packagist/index.ts#L60) function to look for the `metadata-url` property in the response (and the `list` property, to assure that that the repository is not empty and has packages) would fix this issue. I believe the `metadata-url` property can be used just as the `providers-lazy-url` is used now.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by zharinov**
## Context

- Closes: #19495

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Blocked by #20595

**Comment by zharinov** [comment]
The problem is that I can't test it on the real repository, however you can see the call to `packagist.org` is identical to what we do for `v2` in general, once we have `metadata-url`.

**Comment by rarkins** [comment]
I had expected that we could remove the hardcoded logic for packagist.org because the new behavior would be to try v2 API first for all registries anyway. Can that be done, or have I misunderstood, or do you prefer to do it gradual?

**Comment by zharinov** [comment]
Well, I thought about query efficiency, but now I see it's not the case since we cache `getRepositoryMeta()` anyway

**Comment by rarkins** [comment]
1. Try v2 first
2. If it fails, and we can fall back, flag that both in memory as well as a longer term cache (e.g. 24 hours) so that at most we test v2 once per day per registryUrl

BTW it could also be written so that we "test" v2 first before doing an extra query, so that if there's 10 lookups at once then we don't get 10 failures/unnecessary queries

**Comment by zharinov** [comment]
I propose more straightforward yet potentially robust solution: always fetch metadata, cache it for 10 minutes and perform `v2` queries when possible, falling back to old methods otherwise.

File: lib/modules/datasource/packagist/index.ts [hunk 1]
```diff
@@ -115,17 +115,28 @@ export class PackagistDatasource extends Datasource {
...
-    const results = await p.map([pkgUrl, devUrl], (url) =>
-      this.http.getJson(url).then(({ body }) => body)
+  async packagistV2Lookup(
+    registryUrl: string,
+    metadataUrl: string,
+    packageName: string
+  ): Promise<ReleaseResult | null> {
+    let pkgUrl = metadataUrl.replace('%package%', packageName);
+    pkgUrl = parseUrl(pkgUrl) ? pkgUrl : resolveBaseUrl(registryUrl, pkgUrl);
+    const pkgPromise = this.getJson(pkgUrl, z.unknown());
```
Comment (viceice) [inline lib/modules/datasource/packagist/index.ts:129]: the `unknown` is temporary? can we just remove it for now? or is the parameter required?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/datasource/packagist/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/modules/datasource/packagist/index.ts:129]: It's being parsed a bit later, once all requests are performed.
I want to make the Packagist datasource completely schema-driven, but it requires more refactoring PRs in the future

**Review by viceice — APPROVED** [review]

**Pull request merged**